src/dataflow/src/source/loki.rs

```diff
@@ -19,30 +19,83 @@ use crate::source::{SimpleSource, SourceError, Timestamper};
 
 pub struct LokiSourceReader {
     source_id: SourceInstanceId,
-    conn_info: ConnectionInfo,
+    conn_info: LokiConnectionInfo,
     batch_window: Duration,
     query: String,
     client: reqwest::Client,
 }
 
 #[derive(Clone)]
-struct ConnectionInfo {
+pub struct LokiConnectionInfo {
     user: String,
     pw: String,
     endpoint: String,
 }
 
+impl LokiConnectionInfo {
+    /// Loads connection information form the environment. Checks for `LOKI_ADDR`, `LOKI_USERNAME` and `LOKI_PASSWORD`.
+    pub fn from_env() -> LokiConnectionInfo {
+        let mut c = LokiConnectionInfo {
+            user: "".to_string(),
+            pw: "".to_string(),
+            endpoint: "".to_string(),
+        };
+
+        if let Ok(user) = env::var("LOKI_USERNAME") {
+            c.user = user;
+        }
+
+        if let Ok(password) = env::var("LOKI_PASSWORD") {
+            c.pw = password;
+        }
+
+        if let Ok(address) = env::var("LOKI_ADDR") {
+            c.endpoint = address;
+        }
+
+        return c;
+    }
+
+    pub fn with_user(self, user: Option<String>) -> LokiConnectionInfo {
+        if let Some(user) = user {
+            let mut c = self.clone();
+            c.user = user;
+            return c;
+        } else {
+            return self;
+        }
+    }
```

Comment:
```suggestion
    pub fn with_user(mut self, user: Option<String>) -> LokiConnectionInfo {
        if let Some(user) = user {
            self.user = user;
        }
        self
    }
```

(although that'll change if the user/password fields are updated to `Option`s)

Reply:
I didn't want to make it mutable but always return a copy. You know I've written Scala for seven years, right :wink:

Reply:
The `LokiConnectionInfo` is being taken by value anyway (i.e. with a `self` receiver rather than `&self`) so the copy is unnecessary I think!